numpy/ma/core.py

```diff
@@ -201,7 +201,13 @@ def _recursive_fill_value(dtype, f):
     Recursively produce a fill value for `dtype`, calling f on scalar dtypes
     """
     if dtype.names is not None:
-        vals = tuple(_recursive_fill_value(dtype[name], f) for name in dtype.names)
+        # We wrap into `array` here, which ensures we use NumPy cast rules
+        # for integer casts, this allows the use of 99999 as a fill value
+        # for int8.
+        # TODO: This is probably a mess, but should best preserve behavior?
+        vals = tuple(
+                np.array(_recursive_fill_value(dtype[name], f))
+                for name in dtype.names)
```

Comment:
Did you reach the conclusion that this is correct based on a test failure? I don't see an added test, but the lines are not marked by coverage so I assume they are tested.

Reply:
Yes, there were a few test failures (mainly in the recarray code that handles a lot of masked arrays with structs).
They go away with this change, which mirrors the non-struct path (that always wraps into `array()`...

It took me a while to figure that this seems like the easy way out, I don't particularly love it, but it seems to work.  It is a bit unclear if downstream might notice these changes.

Reply:
+1